lib/llvmopencl/Workgroup.cc

```diff
@@ -413,6 +407,77 @@ bool WorkgroupImpl::runOnModule(Module &M, llvm::FunctionAnalysisManager &FAM) {
     F->eraseFromParent();
   }
 
+  // Unify some attributes among all functions; this is necessary because
+  // -O0 compiled builtin library has many attributes setup differently than
+  // -O2 compiled user code, and the inliner will check&refuse to inline.
+  // See hasCompatibleFnAttrs in llvm/include/llvm/IR/Attributes.inc
+  // and getAttributeBasedInliningDecision in llvm/lib/Analysis/InlineCost.cpp
+  std::string TargetCPU;
+  std::string TargetFeatures;
+  for (Function &F : M) {
+    if (TargetCPU.empty() && F.hasFnAttribute("target-cpu")) {
+      TargetCPU = F.getFnAttribute("target-cpu").getValueAsString().str();
+    }
+    if (TargetFeatures.empty() && F.hasFnAttribute("target-features")) {
+      TargetFeatures =
+          F.getFnAttribute("target-features").getValueAsString().str();
+    }
+  }
+
+  for (Function &F : M.functions()) {
+    if (F.isDeclaration())
+      continue;
+
+    // removes optnone/noinline/alwaysinline from F & its callsites
+    markFunctionAlwaysInline(&F);
+    F.removeFnAttr(Attribute::AlwaysInline);
+
+    // these are checked in LLVM's areInlineCompatible -> hasCompatibleFnArgs
+    F.removeFnAttr(Attribute::SafeStack);
+    F.removeFnAttr(Attribute::ShadowCallStack);
+    F.removeFnAttr(Attribute::NoProfile);
+    F.removeFnAttr(Attribute::StackProtect);
+    F.removeFnAttr(Attribute::StackProtectReq);
+    F.removeFnAttr(Attribute::StackProtectStrong);
+    // denorm & strictFP must be compatible
+    // TODO strictfp should depend on program's build opts
+    // denormals are handled by the driver / execution environment
+    F.removeFnAttr(Attribute::StrictFP);
+    F.removeFnAttr("denormal-fp-math");
+    F.removeFnAttr("denormal-fp-math-f32");
+    F.addFnAttr("no-trapping-math", "true");
+    // required because the InlineCost -> areInlineCompatible also calls
+    // TTI->areInlineCompatible; X86TTIImpl::areInlineCompatible checks CPU
+    // features; this might require even more attrs (target-abi?) for RISC-V
+    if (!TargetCPU.empty())
+      F.addFnAttr("target-cpu", TargetCPU);
+    if (!TargetFeatures.empty())
+      F.addFnAttr("target-features", TargetFeatures);
+
+    // the following settings are not necessary for inlining,
+    // but should improve optimization
+    F.removeFnAttr("use-sample-profile");
+    F.removeFnAttr("stack-protector-buffer-size");
+    if (IsDebugEnabled)
+      F.addFnAttr("frame-pointer", "all");
+    else
+      F.removeFnAttr("frame-pointer");
+
+    // these should be safe to enable
+    F.addFnAttr(Attribute::NoBuiltin);
+    F.addFnAttr(Attribute::NoFree);
+    // recursion forbidden by OpenCL
+    F.addFnAttr(Attribute::NoRecurse);
+    // no exceptions or return by callback
+    F.addFnAttr(Attribute::NoCallback);
+    F.addFnAttr(Attribute::NoUnwind);
+#if 0
+    // TBD: find out if infinite recursion is UB in OpenCL.
```

Comment:
Call recursion is not supported in OpenCL C [OpenCL C 6.11.k].

Reply:
Oops, wrong comment. It should be "infinite loop"

Reply:
There doesn't seem to be exceptions in OpenCL C for infinite loops so C99 probably applies here which means they are defined - aka. a kernel would hang if a work-item enters such loop.

Reply:
That's my understanding as well. The 'willreturn" attribute breaks with infinite loops (LLVM replaces them with unreachable inst), so i've fixed the comment.